src/Disks/VolumeJBOD.cpp

```diff
@@ -67,36 +78,66 @@ VolumeJBOD::VolumeJBOD(const VolumeJBOD & volume_jbod,
     : VolumeJBOD(volume_jbod.name, config, config_prefix, disk_selector)
 {
     are_merges_avoided_user_override = volume_jbod.are_merges_avoided_user_override.load(std::memory_order_relaxed);
-    last_used = volume_jbod.last_used.load(std::memory_order_relaxed);
+
+    std::lock_guard lock(volume_jbod.mutex);
+    last_used = volume_jbod.last_used;
 }
 
 DiskPtr VolumeJBOD::getDisk(size_t /* index */) const
 {
-    size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
-    size_t index = start_from % disks.size();
-    return disks[index];
+    std::lock_guard lock(mutex);
+
+    switch (load_balancing)
+    {
+        case VolumeLoadBalancing::ROUND_ROBIN:
+        {
+            size_t start_from = last_used++;
+            size_t index = start_from % disks.size();
+            return disks[index];
+        }
+        case VolumeLoadBalancing::LEAST_USED:
+        {
+            return disks_by_size.top();
+        }
+    }
+    __builtin_unreachable();
 }
 
 ReservationPtr VolumeJBOD::reserve(UInt64 bytes)
 {
+    std::lock_guard lock(mutex);
+
     /// This volume can not store data which size is greater than `max_data_part_size`
     /// to ensure that parts of size greater than that go to another volume(s).
-
     if (max_data_part_size != 0 && bytes > max_data_part_size)
         return {};
 
-    size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
-    size_t disks_num = disks.size();
-    for (size_t i = 0; i < disks_num; ++i)
+    switch (load_balancing)
     {
-        size_t index = (start_from + i) % disks_num;
-
-        auto reservation = disks[index]->reserve(bytes);
-
-        if (reservation)
-            return reservation;
+        case VolumeLoadBalancing::ROUND_ROBIN:
+        {
+            size_t start_from = last_used++;
+            size_t disks_num = disks.size();
+            for (size_t i = 0; i < disks_num; ++i)
+            {
+                size_t index = (start_from + i) % disks_num;
+
+                auto reservation = disks[index]->reserve(bytes);
+
+                if (reservation)
+                    return reservation;
+            }
+            return {};
+        }
+        case VolumeLoadBalancing::LEAST_USED:
+        {
+            DiskPtr disk = disks_by_size.top();
+            disks_by_size.pop();
+            disks_by_size.push(disk);
+            return disk->reserve(bytes);
```

Comment:
this will mess up priority queue, you have just pushed disk with another value of key, reservation should be done before.